packages/0xcert-ethereum-deploy-gateway-contracts/src/contracts/deploy-gateway.sol

```diff
@@ -0,0 +1,391 @@
+  address public assetCreateProxy;
+
+  /**
+   * @dev Mapping of all cancelled deploys.
+   */
+  mapping(bytes32 => bool) public deployCancelled;
+
+  /**
+   * @dev Mapping of all performed deploys.
+   */
+  mapping(bytes32 => bool) public deployPerformed;
+
+  /**
+   * @dev This event emmits when deploy is performed.
+   */
+  event Perform(
+    address indexed _maker,
+    address indexed _taker,
+    address _createdContract,
+    bytes32 _claim
+  );
+
+  /**
+   * @dev This event emmits when deploy is cancelled.
+   */
+  event Cancel(
+    address indexed _maker,
+    address indexed _taker,
+    bytes32 _claim
+  );
+
+  /**
+   * @dev Constructor sets token transfer proxy address.
+   * @param _tokenTransferProxy Address of token transfer proxy.
+   */
+  constructor(
+    address _tokenTransferProxy,
+    address _assetCreateProxy
+  )
+    public
+  {
+    tokenTransferProxy = _tokenTransferProxy;
+    assetCreateProxy = _assetCreateProxy;
+  }
+  
+  /**
+   * @dev Performs the atomic swap that deploys a new Xcert smart contract and at the same time
+   * transfers tokens.
+   * @param _data Data required to make the deploy.
+   * @param _signature Data from the signature.
+   */
+  function perform(
+    DeployData memory _data,
+    SignatureData memory _signature
+  )
+    public
+  {
+    require(_data.taker == msg.sender, TAKER_NOT_EQUAL_TO_SENDER);
+    require(_data.expiration >= now, CLAIM_EXPIRED);
+
+    bytes32 claim = getDeployDataClaim(_data);
+    require(
+      isValidSignature(
+        _data.maker,
+        claim,
+        _signature
+      ),
+      INVALID_SIGNATURE
+    );
+
+    require(!deployCancelled[claim], DEPLOY_CANCELED);
+    require(!deployPerformed[claim], DEPLOY_ALREADY_PERFORMED);
+
+    deployPerformed[claim] = true;
+
+    address xcert = _doActions(_data);
+
+    emit Perform(
+      _data.maker,
+      _data.taker,
+      xcert,
+      claim
+    );
+  }
+
+  /**
+   * @dev @dev Performs the atomic swap that deploys a new Xcert smart contract and at the same time
+   * transfers tokens where performing address does not need to be known before
+   * hand.
+   * @notice When using this function, be aware that the zero address is reserved for replacement
+   * with msg.sender, meaning you cannot send anything to the zero address.
+   * @param _data Data required to make the deploy.
+   * @param _signature Data from the signature.
+   */
+  function performAnyTaker(
+    DeployData memory _data,
+    SignatureData memory _signature
+  )
+    public
+  {
+    require(_data.expiration >= now, CLAIM_EXPIRED);
+
+    bytes32 claim = getDeployDataClaim(_data);
+    require(
+      isValidSignature(
+        _data.maker,
+        claim,
+        _signature
+      ),
+      INVALID_SIGNATURE
+    );
+
+    require(!deployCancelled[claim], DEPLOY_CANCELED);
+    require(!deployPerformed[claim], DEPLOY_ALREADY_PERFORMED);
+
+    deployPerformed[claim] = true;
+
+    address xcert = _doActions(_data);
```

Comment:
Where is the step to substitute 0x0 addresses?

Reply:
When we were discussing `DeployGateway` the use cases for it became quite different then for `OrderGateway` since in this case you don't do any trades but you just delegate deployment to someone else in return for tokens. From this perspective you should know who you delegate to in terms of who the receiver of the tokens will be. But any taker option is still needed so that the delegated party can optimise the deployment process. To keep the same consistency with `OrderGateway` there are two functions for this but the only difference is checking who the taker is. 

Do you see a use case for this? If so I will add address change for token transfer.

Reply:
I added the dynamic receiver. Easier to add it now even tho we don't see a use case than after.

Reply:
That's a fair explanation and if the feature will not be there a note could be added in the .sol.

But you added it and that's cool too. I may be useful in the future. Especially if OrderGateway and DeployGateway ever get merged... or if DeployGateway gets more features. This is not such a ridiculous idea if you look at how tokens-owning-tokens is implemented in https://nft.life (tl;dr it is constantly deploying contracts and assigning them).